File: exclusive-exec-prototype-example/src/main/java/io/averkhoglyad/proto/example/ProtoApplication.java
package io.averkhoglyad.proto.example;

import io.averkhoglyad.proto.exclusive.context.EnableExclusiveExecution;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;

import java.util.stream.IntStream;

@SpringBootApplication
@EnableExclusiveExecution
public class ProtoApplication {

    @Bean
    InheritedComponent someComponent() {
        return new InheritedComponent();
    }

    public static void main(String[] args) {
        var context = SpringApplication.run(ProtoApplication.class, args);

        var bean = context.getBean(InheritedComponent.class);

        doInParallel(() -> bean.test0());
        doInParallel(() -> bean.test1_2(1, 2));
        doInParallel(() -> bean.test1_2(1, 2));
        doInParallel(() -> bean.test1_2(1, 1));
        doInParallel(() -> bean.test1_2(1, -1));
        doInParallel(() -> bean.test1_2(2, 2));
        doInParallel(() -> bean.test1_2(2, 2));
        doInParallel(() -> bean.test1_2(2, 2));

        doInParallel(() -> bean.test3(1, 2, 3));
        doInParallel(() -> bean.test3(1, 2, 3));
        doInParallel(() -> bean.test3(1, 2, 3));

        IntStream.range(0, 3)
                .forEach(i -> doInParallel(() -> bean.test1_2(1, i)));
    }

    private static void doInParallel(Runnable fn) {
        new Thread(() -> {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            fn.run();
        }).start();
    }
}

File: src/main/kotlin/io/averkhoglyad/proto/ProtoApplication.kt
package io.averkhoglyad.proto

import io.averkhoglyad.proto.exclusive.ExclusiveExecutionAspect
import io.averkhoglyad.proto.exclusive.ExclusiveExecutionMethodInterceptor
import io.averkhoglyad.proto.exclusive.ExclusiveExecutionPointcut
import org.springframework.aop.support.DefaultPointcutAdvisor
import org.springframework.beans.factory.BeanFactory
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.context.annotation.Bean
import java.util.stream.IntStream
import kotlin.reflect.KClass

@SpringBootApplication
class ProtoApplication {

//    @Bean
//    fun exclusiveExecutionAspect(): ExclusiveExecutionAspect {
//        return ExclusiveExecutionAspect()
//    }

    @Bean
    fun exclusiveExecutionPointcut(): ExclusiveExecutionPointcut {
        return ExclusiveExecutionPointcut("io.averkhoglyad.proto")
    }

    @Bean
    fun exclusiveExecutionMethodInterceptor(): ExclusiveExecutionMethodInterceptor {
        return ExclusiveExecutionMethodInterceptor()
    }

    @Bean
    fun defaultPointcutAdvisor(): DefaultPointcutAdvisor {
        return DefaultPointcutAdvisor(exclusiveExecutionPointcut(), exclusiveExecutionMethodInterceptor())
    }

    @Bean
    fun someComponent(): SomeComponent {
        return SomeComponent()
    }
}

fun main(vararg args: String) {
    val context = runApplication<ProtoApplication>(*args)

    val bean = context.getBean<SomeComponent>()

    doInParallel { bean.test0() }
    doInParallel { bean.test1_2(1, 2) }
    doInParallel { bean.test1_2(1, 2) }
    doInParallel { bean.test1_2(2, 2) }

    IntStream.range(0, 3)
        .parallel()
        .forEach { _ -> bean.test1_2(1, 2) }
}

private fun doInParallel(fn: () -> Unit) {
    Thread {
        try {
            Thread.sleep(1000)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
        fn()
    }.start()
}

inline fun <reified T : Any> runApplication(vararg args: String): ConfigurableApplicationContext = SpringApplication.run(T::class.java, *args)

inline fun <reified E: Any> BeanFactory.getBean() = getBean(E::class)

inline fun <E : Any> BeanFactory.getBean(klass: KClass<E>) = getBean(klass.java)
